store/tikv/2pc_exec.go

```diff
@@ -0,0 +1,535 @@
+	c.filter = func(key []byte) bool {
+		return bytes.Compare(key, start) < 0 || bytes.Compare(key, end) > 0
+	}
+	return batch.mutations.len(), nil
+}
+
+func (c *execController) handleMutations() (err error) {
+	collector := c.getCollector(nil, nil)
+	for {
+		var batch *batchMutations
+		batch, err = collector.Collect()
+		if err != nil {
+			c.handleError(err)
+			break
+		}
+		if batch == nil {
+			break
+		}
+
+		if collector.Finished() {
+			bo, cancel := c.bo.Fork()
+			job := execJob{batch, bo, cancel}
+			c.handleSingleBatch(&job)
+			break
+		}
+
+		if err = c.dispatchBatch(batch); err != nil {
+			break
+		}
+	}
+	c.Wait()
+	if err == nil {
+		select {
+		case err = <-c.errCh:
+		default:
+		}
+	}
+	close(c.jobCh)
+
+	return errors.Trace(err)
+}
+
+func (c *execController) containsPrimaryKey() bool {
+	it := c.mutations.Iter(c.committer.primaryKey, nil)
+	m := it.Next()
+	return bytes.Equal(m.key, c.committer.primaryKey)
+}
+
+func (c *execController) getCollector(start, end []byte) *mutationBatchCollector {
+	_, actionIsPrewrite := c.action.(actionPrewrite)
+	it := c.mutations.Iter(start, end)
+	if c.filter != nil {
+		it.WithFilter(c.filter)
+	}
+
+	sizer := func(m *mutation) int { return len(m.key) }
+	if actionIsPrewrite {
+		sizer = func(m *mutation) int { return len(m.key) + len(m.value) }
+	}
+
+	return &mutationBatchCollector{
+		src:        c.committer.mapWithRegion(c.bo, it),
+		sizer:      sizer,
+		limit:      txnCommitBatchSize,
+		primaryKey: c.committer.primaryKey,
+		collector:  c.action.collectMutation,
+	}
+}
+
+func (c *execController) dispatchBatch(batch *batchMutations) error {
+	bo, cancel := c.bo.Fork()
+	job := execJob{batch, bo, cancel}
+	c.Add(1)
+
+	select {
+	case c.jobCh <- job:
+		return nil
+	default:
+		c.tryAddWorker()
+	}
+
+	if c.cancel == nil {
+		c.jobCh <- job
+		return nil
+	}
+
+	// We can cancel dispatcher by return error.
+	select {
+	case c.jobCh <- job:
+		return nil
+	case err := <-c.errCh:
+		c.Done()
+		return err
+	}
+}
+
+func (c *execController) workerLoop() {
+	for job := range c.jobCh {
+		if !c.terminated() {
+			c.handleSingleBatch(&job)
+		}
+		c.Done()
+	}
+}
+
+func (c *execController) handleSingleBatch(job *execJob) {
+	beforeSleep := job.bo.totalSleep
+	defer c.updateSingleBatchDetail(job.bo, beforeSleep)
+
+	retry, err := c.action.handleSingleBatch(c.committer, job.bo, job.batch)
+	if err == nil && !retry {
+		job.cancel()
+		return
+	}
+	if !retry {
+		c.handleError(err)
+		job.cancel()
+		return
+	}
+	c.handleRegionError(err, job)
+}
+
+func (c *execController) handleRegionError(retryErr error, retryJob *execJob) {
+	keys := retryJob.batch.mutations.keys
+	start := kv.Key(keys[0])
+	end := kv.Key(keys[len(keys)-1]).Next()
+
+	for {
+		if retryErr != nil {
+			err := retryJob.bo.Backoff(BoRegionMiss, retryErr)
+			if err != nil {
+				c.handleError(err)
+				return
+			}
+		}
+
+		collector := c.getCollector(start, end)
+		for {
```

Comment:
This inner loop seems to repeat some code of `handleMutations`?

Reply:
They have a little difference which cannot be extracted to a function easily.